### PR TITLE
feat: Add scalar storage support for DynamoDB 

### PR DESCRIPTION
**Pull request by gauthamchandra**
This change allows GPTCache to use DynamoDB as the underlying scalar
storage for the cache. Addresses #200 

The underlying implementation uses 2 tables:

- `gptcache_questions` - which holds all questions and session
  information.
- `gptcache_reports` - which holds the reporting information.

Normally, we would do a single table design and rollup
`gptcache_reports` into the same table as `gptcache_questions`. However,
this was not done for one key reason: billing.

In the event a lot of analytics data is being created, then table scans
for operations like `count()` and `get_ids()` would also involve reading
these reporting rows before filtering them out, resulting in higher
read costs for end users of GPTCache.

#### `gpt_cache` table

image

#### Secondary Indexes:

**gsi_items_by_type:**

image

**gsi_questions_by_deletion_status**:
image

### CODE REVIEW

File: tests/requirements.txt [hunk 1]
```diff
@@ -15,6 +15,7 @@ pytest-html==3.1.1
 pytest-sugar==0.9.5
 pytest-parallel
 psycopg2-binary
+testcontainers==3.7.1
```
Comment (gauthamchandra) [inline tests/requirements.txt:18]: Woops. This was added accidentally. Stripping.

**Comment by gauthamchandra** [comment]
Sorry for the failing tests. Turns out that specifying the `region_name` for AWS overrides the `endpoint_url` field so it ended up trying to connect directly to `us-east-2` as part of the tests instead of the local DynamoDB container.

My latest changes should hopefully fix the issue. 🙏🏼

**Comment by gauthamchandra** [comment]
It seems a totally unrelated test is failing. Might be finicky 🤔

**Comment by SimFG** [comment]
@gauthamchandra thanks your patience, and i will fix it in the next pr

**Comment by mikeblackk** [comment]
Has this been released? can we use dynamodb now as a storage for caching?

**Comment by SimFG** [comment]
yes, you can. the usage is like other cache storage, and its params include:
```
elif name == "dynamo":
            from gptcache.manager.scalar_data.dynamo_storage import DynamoStorage

            return DynamoStorage(
                aws_access_key_id=kwargs.get("aws_access_key_id"),
                aws_secret_access_key=kwargs.get("aws_secret_access_key"),
                aws_region_name=kwargs.get("region_name"),
                aws_profile_name=kwargs.get("aws_profile_name"),
                aws_endpoint_url=kwargs.get("endpoint_url"),
```

**Comment by gauthamchandra** [comment]
@mikeblackk, just want to clarify that it uses `boto3` under the hood so not all params (like `aws_endpoint_url`, etc.) are required.

As a result, it relies on the standard AWS resolution logic (so if you have ENV vars like `AWS_ACCESS_KEY_ID`, then it will be automatically picked up without explicitly specifying). The parameters like `aws_endpoint_url` act as overrides so that you have an easier time developing on your local machine where you might be using LocalStack or have multiple AWS profiles.

Let me know if you run into any issues. I am not an expert on Dynamo by any means so there might be issues which need to be fixed 😄

**Pull request merged**